compiler/cpp/src/thrift/generate/t_cpp_generator.cc

```diff
@@ -1117,7 +1129,7 @@ void t_cpp_generator::generate_struct_declaration(ofstream& out,
         << " &b);" << endl << endl;
   }
 
-  if (is_user_struct) {
+  if (is_user_struct && !has_custom_ostream(tstruct)) {
```

Comment:
This needs to change; on ostream declaration must be in the header even if thrift does not emit an implementation for it so that a thrift class that emits output code can contain a thrift class that does not.